beacon-chain/db/kv/state.go

```diff
@@ -686,84 +670,19 @@ func (s *Store) slotByBlockRoot(ctx context.Context, tx *bolt.Tx, blockRoot []by
-			if root == nil {
-				continue
-			}
-			if key >= slot {
-				break
-			}
-			best = root
-		}
-		return nil
-	}); err != nil {
-		return nil, err
-	}
-
-	var st state.ReadOnlyBeaconState
-	var err error
-	if best != nil {
-		st, err = s.State(ctx, bytesutil.ToBytes32(best))
-		if err != nil {
-			return nil, err
-		}
-	}
-	if st == nil || st.IsNil() {
-		st, err = s.GenesisState(ctx)
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	return []state.ReadOnlyBeaconState{st}, nil
-}
-
-// createStateIndicesFromStateSlot takes in a state slot and returns
-// a map of bolt DB index buckets corresponding to each particular key for indices for
-// data, such as (shard indices bucket -> shard 5).
-func createStateIndicesFromStateSlot(ctx context.Context, slot types.Slot) map[string][]byte {
-	ctx, span := trace.StartSpan(ctx, "BeaconDB.createStateIndicesFromState")
-	defer span.End()
-	indicesByBucket := make(map[string][]byte)
-	// Every index has a unique bucket for fast, binary-search
-	// range scans for filtering across keys.
-	buckets := [][]byte{
-		stateSlotIndicesBucket,
-	}
-
-	indices := [][]byte{
-		bytesutil.SlotToBytesBigEndian(slot),
-	}
-	for i := 0; i < len(buckets); i++ {
-		indicesByBucket[string(buckets[i])] = indices[i]
-	}
-	return indicesByBucket
-}
-
-// CleanUpDirtyStates removes states in DB that falls to under archived point interval rules.
-// Only following states would be kept:
-// 1.) state_slot % archived_interval == 0. (e.g. archived_interval=2048, states with slot 2048, 4096... etc)
-// 2.) archived_interval - archived_interval/3 < state_slot % archived_interval
-//   (e.g. archived_interval=2048, states with slots after 1365).
-//   This is to tolerate skip slots. Not every state lays on the boundary.
-// 3.) state with current finalized root
-// 4.) unfinalized States
+var errSavedStateMissingBlock = errors.New("Could not find block corresponding to saved state")
+
+// CleanUpDirtyStates attempts to maintain the promise to save approximately <head slot / save state interval> states.
+// To do that, we save about 1 state every eg 2048 slots (default slotsPerArchivedPoint value), calling the slot
+// where the save happened the "save point". Due to skipped slots, there may not be a block at a multiple of 2048,
+// in which case the saved state point will be at the slot where the last block was previously included in the interval.
+// We don't want to delete the most recently finalized state, which is saved to the same database,
+// and in long periods of non-finality, stategen may also write a state every 128 slots to aid in recovery.
+// So we preserve:
+//   1. any state where the slot number is a multiple of 2048 (slot % 2048 == 0)
+//   2. any state with a slot number within 682 slots (2048/3) of a such a save point,
+//   3. most recently finalized state
+//   4. non-finalized states used by stategen
 func (s *Store) CleanUpDirtyStates(ctx context.Context, slotsPerArchivedPoint types.Slot) error {
```

Comment:
This is the main thing I need feedback on, because the rest of the PR doesn't make sense unless this function works without the state slot index. The previous implementation would definitely run faster, because this implementation has to read more data and decompress it, but this removes overhead spent on maintaining the index throughout the life of the process. We're trading more work when the node starts up for less work the rest of the time. Notable that when we move to persistence based around finalization assumptions (for validator delta tree), this code won't be needed anymore.